File: client/src/main/java/se/kth/id2203/kvstore/ClientService.java
/*
 * The MIT License
 *
 * Copyright 2017 Lars Kroll <lkroll@kth.se>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package se.kth.id2203.kvstore;

import com.google.common.base.Optional;
import com.google.common.util.concurrent.SettableFuture;
import java.util.Map;
import java.util.TreeMap;
import java.util.UUID;
import java.util.concurrent.Future;
import org.slf4j.LoggerFactory;
import se.kth.id2203.networking.Message;
import se.kth.id2203.networking.NetAddress;
import se.kth.id2203.overlay.Connect;
import se.kth.id2203.overlay.RouteMsg;
import se.sics.kompics.ClassMatchedHandler;
import se.sics.kompics.ComponentDefinition;
import se.sics.kompics.Handler;
import se.sics.kompics.Kompics;
import se.sics.kompics.KompicsEvent;
import se.sics.kompics.Positive;
import se.sics.kompics.Start;
import se.sics.kompics.network.Network;
import se.sics.kompics.timer.ScheduleTimeout;
import se.sics.kompics.timer.Timeout;
import se.sics.kompics.timer.Timer;

/**
 *
 * @author Lars Kroll <lkroll@kth.se>
 */
public class ClientService extends ComponentDefinition {
    
    private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(ClientService.class);
    //******* Ports ******
    final Positive<Timer> timer = requires(Timer.class);
    final Positive<Network> net = requires(Network.class);
    //******* Fields ******
    private final NetAddress self = config().getValue("id2203.project.address", NetAddress.class);
    private final NetAddress server = config().getValue("id2203.project.bootstrap-address", NetAddress.class);
    private Optional<Connect.Ack> connected = Optional.absent();
    private final Map<UUID, SettableFuture<OpResponse>> pending = new TreeMap<>();

    //******* Handlers ******
    protected final Handler<Start> startHandler = new Handler<Start>() {
        
        @Override
        public void handle(Start event) {
            LOG.debug("Starting client on {}. Waiting to connect...", self);
            long timeout = (config().getValue("id2203.project.keepAlivePeriod", Long.class) * 2);
            ScheduleTimeout st = new ScheduleTimeout(timeout);
            st.setTimeoutEvent(new ConnectTimeout(st));
            trigger(new Message(self, server, new Connect(st.getTimeoutEvent().getTimeoutId())), net);
            trigger(st, timer);
        }
    };
    protected final ClassMatchedHandler<Connect.Ack, Message> connectHandler = new ClassMatchedHandler<Connect.Ack, Message>() {
        
        @Override
        public void handle(Connect.Ack content, Message context) {
            LOG.info("Client connected to {}, cluster size is {}", server, content.clusterSize);
            connected = Optional.of(content);
            Console c = new Console(ClientService.this);
            Thread tc = new Thread(c);
            tc.start();
        }
    };
    protected final Handler<ConnectTimeout> timeoutHandler = new Handler<ConnectTimeout>() {
        
        @Override
        public void handle(ConnectTimeout event) {
            if (!connected.isPresent()) {
                LOG.error("Connection to server {} did not succeed. Shutting down...", server);
                Kompics.asyncShutdown();
            } else {
                Connect.Ack cack = connected.get();
                if (!cack.id.equals(event.getTimeoutId())) {
                    LOG.error("Received wrong response id earlier! System may be inconsistent. Shutting down...", server);
                    System.exit(1);
                }
            }
        }
    };
    protected final Handler<OpWithFuture> opHandler = new Handler<OpWithFuture>() {
        
        @Override
        public void handle(OpWithFuture event) {
            RouteMsg rm = new RouteMsg(event.op.key, event.op); // don't know which partition is responsible, so ask the bootstrap server to forward it
            trigger(new Message(self, server, rm), net);
            pending.put(event.op.id, event.f);
        }
    };
    protected final ClassMatchedHandler<OpResponse, Message> responseHandler = new ClassMatchedHandler<OpResponse, Message>() {
        
        @Override
        public void handle(OpResponse content, Message context) {
            LOG.debug("Got OpResponse: {}", content);
            SettableFuture<OpResponse> sf = pending.remove(content.id);
            if (sf != null) {
                sf.set(content);
            } else {
                LOG.warn("ID {} was not pending! Ignoring response.", content.id);
            }
        }
    };
    
    {
        subscribe(startHandler, control);
        subscribe(timeoutHandler, timer);
        subscribe(connectHandler, net);
        subscribe(opHandler, loopback);
        subscribe(responseHandler, net);
    }
    
    Future<OpResponse> op(String key) {
        Operation op = new Operation(key);
        OpWithFuture owf = new OpWithFuture(op);
        trigger(owf, onSelf);
        return owf.f;
    }
    
    public static class OpWithFuture implements KompicsEvent {
        
        public final Operation op;
        public final SettableFuture<OpResponse> f;
        
        public OpWithFuture(Operation op) {
            this.op = op;
            this.f = SettableFuture.create();
        }
    }
    
    public static class ConnectTimeout extends Timeout {
        
        ConnectTimeout(ScheduleTimeout st) {
            super(st);
        }
    }
}

File: client/src/main/scala-2.11/kvstore/ClientService.scala
package kvstore

import java.util.UUID

import com.google.common.util.concurrent.SettableFuture
import com.typesafe.scalalogging.StrictLogging
import se.sics.kompics.network.Network
import se.sics.kompics.sl._
import se.sics.kompics.timer.{SchedulePeriodicTimeout, Timeout, Timer}
import se.sics.kompics.{Kompics, KompicsEvent, Start}
import stormy.kv._
import stormy.networking.{NetAddress, NetMessage}
import stormy.overlay.{Ack, Connect, RouteMessage}


class ClientService extends ComponentDefinition with StrictLogging {

    val self: NetAddress = config.getValue("stormy.address", classOf[NetAddress])
    val coordinator: NetAddress = config.getValue("stormy.coordinatorAddress", classOf[NetAddress])
    private val pending = new java.util.TreeMap[String, SettableFuture[OperationResponse]]()
    var network: PositivePort[Network] = requires[Network]
    var timer: PositivePort[Timer] = requires[Timer]
    private var connected: Option[Ack] = None


    ctrl uponEvent {
        case _: Start => handle {
            logger.info(s"Starting client at $self")
            val timeout: Long = cfg.getValue[Long]("stormy.keepAlivePeriod") * 2
            val spt = new SchedulePeriodicTimeout(timeout, timeout)
            spt.setTimeoutEvent(ConnectTimeout(spt))
            logger.debug(s"Setting up with timeout ID ${spt.getTimeoutEvent.getTimeoutId}")
            trigger(NetMessage(self, coordinator, Connect(spt.getTimeoutEvent.getTimeoutId.toString)) -> network)
            trigger(spt -> timer)
        }
    }

    loopbck uponEvent {
        case ctx@OpWithFuture(op: Operation) => handle {
            val msg: RouteMessage = RouteMessage(op.key, op) // don't know which partition is responsible, so ask the bootstrap server to forward it
            trigger(NetMessage(self, coordinator, msg) -> network)
            pending.put(op.id, ctx.sf)
        }
    }

    network uponEvent {
        case NetMessage(source, self, ack: Ack) => handle {
            logger.info(s"Client connected to $source, cluster size is ${ack.clusterSize}")
            connected = Some(ack)
            val console: Console = new Console(ClientService.this)
            val th: Thread = new Thread(console)
            th.start()
        }
        case NetMessage(source, self, response: OperationResponse) => handle {
            logger.debug(s"Received response $response")
            val sf: Option[SettableFuture[OperationResponse]] = Some(pending.remove(response.id))
            sf match {
                // TODO Confirm majority in coordinator's Router to avoid these checks
                case None =>        // Ignore
                case Some(null) =>  // Ignore
                case Some(value) => value.set(response)
            }
        }
    }

    timer uponEvent {
        case ev: ConnectTimeout => handle {
            logger.debug(s"Received event: $ev\n${connected.get}\n${connected.get.id}\n${ev.getTimeoutId}")
            connected match {
                case Some(ack: Ack) =>
                    if (!ack.id.equals(ev.getTimeoutId.toString)) {
                        logger.error("Received wrong response ID earlier! System may be inconsistent. Shutting down")
                        System.exit(1)
                    }
                case None =>
                    logger.error(s"Connection to server $coordinator did not succeed. Shutting down.")
                    Kompics.asyncShutdown()
            }
        }
    }

    private[kvstore] def op(key: String) = {
        val op = GetOperation(key, UUID.randomUUID().toString, self)
        val owf = OpWithFuture(op)
        trigger(owf, onSelf)
        owf.sf
    }
    private[kvstore] def op(cmdline: Array[String]) = {
        val key = cmdline(1)
        var op: Option[Operation] = None
        cmdline(0).toLowerCase match {
            case "status" =>
                op = Some(StatusRequest(key, UUID.randomUUID().toString, self))
            case "put" =>
                val value = cmdline(2)
                op = Some(PutOperation(key, value, UUID.randomUUID().toString, self))
            case "get" =>
                op = Some(GetOperation(key, UUID.randomUUID().toString, self))
            case "cas" =>
                val refValue = cmdline(2)
                val newValue = cmdline(3)
                op = Some(CASOperation(key, refValue, newValue, UUID.randomUUID().toString, self))

        }
        val owf = OpWithFuture(op.get)
        trigger(owf, onSelf)
        owf.sf
    }


}

case class ConnectTimeout(spt: SchedulePeriodicTimeout) extends Timeout(spt)

case class OpWithFuture(op: Operation) extends KompicsEvent {
    val sf: SettableFuture[OperationResponse] = SettableFuture.create[OperationResponse]
}